raster-foundry/granary
# Add prediction endpoints

## Overview

This PR adds prediction endpoints. Predictions are connected to models, can be searched by model ID and status (or by nothing), and can be fetched by id. They cannot be updated directly or deleted.

### Checklist

- [x] Any new migrations have been audited to make sure they won't kill the application while being applied
- [x] Any new API endpoints have tests

### Notes

Things got a little more exciting in `tapir`-land here. The `list` endpoint for predictions has _two_ parameters, instead of one, which required the `Function.tupled` thing in `.toRoutes(logic)`. Also, the `create` endpoint had _two types_ of error output, which was a fun puzzle -- if you don't tell `oneOf` what to expect of the different branches, it will assume you want a `Product with Serializable`. To get around this, I tried a `Coproduct` strategy, before making the `CrudError` case classes into an ADT. Puzzles!

There aren't currently tests for filtering by status for predictions. This is deliberate, since there's currently no way for a prediction to change status. That should happen in #25 

## Testing Instructions

- tests
- try to create a prediction for the water model:

```bash
$ echo '{"modelId": "b65116e0-60a5-4979-86bd-f40b54a1cb6f", "arguments": {"RED_BAND": "s3://foo/bar.tif", "GREEN_BAND": "s3://foo/bar.tif", "OUTPUT_LOCATION": "s3://foo/bar.geojson"}}' | http :8081/api/predictions
```

- it should succeed
- remove or misspell a parameter in `arguments` of your choosing and repeat the POST
- it should fail

Closes #13
